Detectors/TPC/reconstruction/include/TPCReconstruction/HwClusterer.h

```diff
@@ -15,135 +15,167 @@
-    int     mPadsPerCF;
-    int     mTimebinsPerCF;
-    int     mCfPerRow;
-    int     mLastTimebin;
-
-    std::vector<o2::TPC::HwCluster>* mClusterArray;    ///< Internal cluster storage
-    
-    CalDet<float>* mNoiseObject;
-    CalDet<float>* mPedestalObject;
+
+    /// Handling of the parallel cluster finder threads
+    /// \param iTimeBinMin Minimum time bin to be processed
+    /// \param iTimeBinMax Maximum time bin to be processed
+    /// \param mcDigitTruth MC Truth information associated with the digits
+    /// \param eventCount Event counter
+    void ProcessTimeBins(int iTimeBinMin, int iTimeBinMax,
+        MCLabelContainer const* mcDigitTruth, int eventCount);
+
+    /*
+     * class members
+     */
+    bool mAssignChargeUnique;               ///< Setting for CF to use charge only for one cluster
+    bool mEnableNoiseSim;                   ///< Switch for noise simulation
+    bool mEnablePedestalSubtraction;        ///< Switch for pedestal subtraction
+    bool mIsContinuousReadout;              ///< Switch for continuous readout
+    int mLastTimebin;                       ///< Last time bin of previous event
+    unsigned mCRUMin;                       ///< Minimum CRU ID to be processed
+    unsigned mCRUMax;                       ///< Maximum CRU ID to be processed
+    unsigned mPadsPerCF;                    ///< Number of pads per cluster finder instance
+    unsigned mTimebinsPerCF;                ///< Number of time bins per cluster finder instance
+    unsigned mNumThreads;                   ///< Number of parallel processing threads
+    float mMinQDiff;                        ///< Minimum charge difference between neighboring pads / time bins
+
+    std::vector<std::vector<std::vector<std::shared_ptr<HwClusterFinder>>>> mClusterFinder;     ///< Cluster finder container for each row in each CRU
+    std::vector<std::vector<std::vector<std::tuple<Digit const*, int, int>>>> mDigitContainer;  ///< Sorted digit container for each row in each CRU. Tuple consists of pointer to digit, original digit index and event count
+
+    std::vector<std::vector<Cluster>> mClusterStorage;                                          ///< Cluster storage for each CRU
+    std::vector<std::vector<std::vector<std::pair<int,int>>>> mClusterDigitIndexStorage;        ///< Container for digit indices, used in found clusters. Pair consists of original digit index and event count
+
+    std::vector<Cluster>* mClusterArray;        ///< Pointer to output cluster storage
+    MCLabelContainer* mClusterMcLabelArray;     ///< Pointer to MC Label storage
+
+    std::shared_ptr<CalDet<float>> mNoiseObject;                ///< Pointer to the CalDet object for noise simulation
+    std::shared_ptr<CalDet<float>> mPedestalObject;             ///< Pointer to the CalDet object for the pedestal subtraction
+
+    std::map<int, std::unique_ptr<MCLabelContainer>> mLastMcDigitTruth; ///< Buffer for digit MC truth information
```

Comment:
is this some kind of FIFO, or what does the `Last` mean? If I understand correctly, this is supposed to keep a set of containers for multiple events, correct? I don't see it used in the code though. How does this relate to the above defined mClusterMcLabelArray?

Reply:
Yes, it is a kind of FIFO and is used only in HwClusterer.cxx line 460.

It stores the MC Truth information of the last event(s) because the actual ClusterFinder doesn't work only on digits of the "current" event. The digits there are kept also across event-boundaries to enable the continuous simulation.

During this, I also found, that there are some events in the simulation consisting of only very few timebins (even only a single timebin). That is, why it is not enough to store only the information from the "last" event, but from several previous events.

Reply:
Is anything missing to accept this pull request?